src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp

```diff
@@ -46,13 +46,22 @@ class G1ResetSkipCompactingClosure : public HeapRegionClosure {
     uint region_index = r->hrm_index();
     // Only for skip-compaction regions; early return otherwise.
     if (!_collector->is_skip_compacting(region_index)) {
-
       return false;
     }
-    assert(_collector->live_words(region_index) > _collector->scope()->region_compaction_threshold() ||
-         !r->is_starts_humongous() ||
-         _collector->mark_bitmap()->is_marked(cast_to_oop(r->bottom())),
-         "must be, otherwise reclaimed earlier");
+#ifdef ASSERT
+    if (r->is_humongous()) {
+      oop obj = cast_to_oop(r->humongous_start_region()->bottom());
+      assert(_collector->mark_bitmap()->is_marked(obj), "must be live");
+    } else if (r->is_open_archive()) {
+      bool is_empty = (_collector->live_words(r->hrm_index()) == 0);
+      assert(!is_empty, "should contain at least one live obj");
+    } else if (r->is_closed_archive()) {
+      // nothing to assert
+    } else {
+      assert(_collector->live_words(region_index) > _collector->scope()->region_compaction_threshold(),
+             "should be quite full");
+    }
+#endif
```

Comment:
What do you think about extracting this to a function. Something like:
```
void assert_should_reset(HeapRegion*) PRODUCT_RETURN;
```
But you might be able to come up with a better name :)

Reply:
Considering this method is fairly short, I think having those asserts inlined doesn't hinder readability; instead, provides some documentation concerning in what context this method is invoked.

Reply:
I think the fact that the method is short a good reason to extract the asserts, now the bulk of the method is asserts. If we want more documentation I would rather see a comment explaining in what situations regions might be marked as "skip compacting". We do have some documentation on this in `G1FullGCHeapRegionAttr`.

I'm fine with leaving the asserts here if you and others prefer it. One other question, can we ever end up here with a closed archive region? Those are marked `SkipMarking` so I guess they should be filtered out above?

Reply:
> If we want more documentation I would rather see a comment explaining in what situations regions might be marked as "skip compacting".

These asserts serve as "runnable" documentation -- less likely to deviate from the actual code.

> One other question, can we ever end up here with a closed archive region? Those are marked SkipMarking so I guess they should be filtered out above?

Yes; you are right. I should put `ShouldNotReachHere()` for closed-archive regions.